// @absmartly/javascript-sdk/es/index.js
"use strict";

var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SDK = exports.ContextPublisher = exports.ContextDataProvider = exports.Context = exports.AbortController = exports.mergeConfig = void 0;
const context_1 = __importDefault(require("./context"));
exports.Context = context_1.default;
const sdk_1 = __importDefault(require("./sdk"));
exports.SDK = sdk_1.default;
const config_1 = require("./config");
Object.defineProperty(exports, "mergeConfig", {
  enumerable: true,
  get: function () {
    return config_1.mergeConfig;
  }
});
const provider_1 = require("./provider");
Object.defineProperty(exports, "ContextDataProvider", {
  enumerable: true,
  get: function () {
    return provider_1.ContextDataProvider;
  }
});
const publisher_1 = require("./publisher");
Object.defineProperty(exports, "ContextPublisher", {
  enumerable: true,
  get: function () {
    return publisher_1.ContextPublisher;
  }
});
const abort_1 = require("./abort");
Object.defineProperty(exports, "AbortController", {
  enumerable: true,
  get: function () {
    return abort_1.AbortController;
  }
});
exports.default = {
  mergeConfig: config_1.mergeConfig,
  AbortController: abort_1.AbortController,
  Context: context_1.default,
  ContextDataProvider: provider_1.ContextDataProvider,
  ContextPublisher: publisher_1.ContextPublisher,
  SDK: sdk_1.default
};
